### ISSUE TITLE
Question - OPT 175B serving - Consolidated shards

### ISSUE THREAD

**Issue by BalajiAJ** [alpa-projects/alpa#554]
Hi,

We are following this guide [OPT-175B using Alpa](https://alpa-projects.github.io/tutorials/opt_serving.html) to host OPT 175B model, using Alpa and had below questions:

1. We are referring to this [folder](https://github.com/alpa-projects/alpa/tree/main/examples/opt_serving) in Alpa GitHub repo, please let us know if this is the right place for setup.
2. Can we directly use the 992 files (shards) released by metaseq for OPT-175B or should we first consolidate the shards to 8, as suggested in [instructions](https://github.com/facebookresearch/metaseq/blob/main/projects/OPT/download_opt175b.md) then run the Alpa weights converter script.

Please advise. Thank you!

**Comment by zhisbug** [comment]
@BalajiAJ Thank you for your interest in Alpa.

1. Yes it is.
2. The procedures are 
 (a) consolidate it into 1 file 
 (b) call [this script](https://github.com/alpa-projects/alpa/blob/main/examples/opt_serving/scripts/convert_to_numpy_weights.py) to save them by the weight name in NumPy formats. 

You can download the [2.7B weight](https://drive.google.com/file/d/1ayIaKRhxF9osZWgcFG-3vSkjcepSWdQd/view?usp=sharing) and see an example format on small models.
We're working on providing an optimized version of the script used in for 2(a), 

@BalajiAJ If you have trouble doing 2(a) or you need to serve it immediately, you can join our slack and we can help on this.

**Comment by BalajiAJ** [comment]
> @BalajiAJ Thank you for your interest in Alpa.
> 
> 1. Yes it is.
> 2. The procedures are
>    (a) consolidate them into 1 file
>    (b) call [this script](https://github.com/alpa-projects/alpa/blob/main/examples/opt_serving/scripts/convert_to_numpy_weights.py) to save them by the weight name in NumPy formats.
> 
> You can download the [2.7B weight](https://drive.google.com/file/d/1ayIaKRhxF9osZWgcFG-3vSkjcepSWdQd/view?usp=sharing) and see an example format on small models. We're working on providing an optimized version of the script used in for 2(a),
> 
> @BalajiAJ If you have trouble doing 2(a) or you need to serve it immediately, you can join our slack and we can help on this.

@zhisbug Thanks for your guidance, I am facing some issues, below is the summary:

1. On running **convert_to_numpy_weights** script, for the consolidated shard file, we get 1157 numpy weight files. Not sure if the count is less, because while running benchmark_text_gen.py, we are getting errors - no weight files found (stack trace below).

StackTrace:

```
opt/conda/lib/python3.7/site-packages/requests/__init__.py:114: RequestsDependencyWarning: urllib3 (1.26.9) or chardet (5.0.0)/charset_normalizer (2.0.12) doesn't match a supported version!
  RequestsDependencyWarning,
Load model alpa/opt-175b ... (This can take several minutes for very large models)
(raylet) /opt/conda/lib/python3.7/site-packages/requests/__init__.py:114: RequestsDependencyWarning: urllib3 (1.26.9) or chardet (5.0.0)/charset_normalizer (2.0.12) doesn't match a supported version!
(raylet)   RequestsDependencyWarning,
(raylet) /opt/conda/lib/python3.7/site-packages/requests/__init__.py:114: RequestsDependencyWarning: urllib3 (1.26.9) or chardet (5.0.0)/charset_normalizer (2.0.12) doesn't match a supported version!
(raylet)   RequestsDependencyWarning,
(raylet) /opt/conda/lib/python3.7/site-packages/requests/__init__.py:114: RequestsDependencyWarning: urllib3 (1.26.9) or chardet (5.0.0)/charset_normalizer (2.0.12) doesn't match a supported version!
(raylet)   RequestsDependencyWarning,
(raylet) /opt/conda/lib/python3.7/site-packages/requests/__init__.py:114: RequestsDependencyWarning: urllib3 (1.26.9) or chardet (5.0.0)/charset_normalizer (2.0.12) doesn't match a supported version!
(raylet)   RequestsDependencyWarning,
(raylet) /opt/conda/lib/python3.7/site-packages/requests/__init__.py:114: RequestsDependencyWarning: urllib3 (1.26.9) or chardet (5.0.0)/charset_normalizer (2.0.12) doesn't match a supported version!
(raylet)   RequestsDependencyWarning,
(raylet) /opt/conda/lib/python3.7/site-packages/requests/__init__.py:114: RequestsDependencyWarning: urllib3 (1.26.9) or chardet (5.0.0)/charset_normalizer (2.0.12) doesn't match a supported version!
(raylet)   RequestsDependencyWarning,
2022-06-29 18:26:20,794	ERROR worker.py:95 -- Unhandled error (suppress with 'RAY_IGNORE_UNHANDLED_ERRORS=1'): ray::MeshHostWorker.load_opt_params_worker_func() (pid=15310, ip=10.128.0.106, repr=<alpa.device_mesh.MeshHostWorker object at 0x7f4e641bbd90>)
  File "/home/user1/alpa/examples/opt_serving/model/opt_model.py", line 785, in load_opt_params_worker_func
    wq = load_array(load_prefix + "self_attn.q_proj.weight")
  File "/home/user1/alpa/examples/opt_serving/model/opt_model.py", line 748, in load_array
    return np.load(os.path.join(path, key))
  File "/opt/conda/lib/python3.7/site-packages/numpy/lib/npyio.py", line 417, in load
    fid = stack.enter_context(open(os_fspath(file), "rb"))
FileNotFoundError: [Errno 2] No such file or directory: '/home/user1/alpa/examples/opt_serving/model/175B_resharded/numpy_weights/decoder.layers.48.self_attn.q_proj.weight'
2022-06-29 18:26:20,862	ERROR worker.py:95 -- Unhandled error (suppress with 'RAY_IGNORE_UNHANDLED_ERRORS=1'): ray::MeshHostWorker.load_opt_params_worker_func() (pid=15307, ip=10.128.0.106, repr=<alpa.device_mesh.MeshHostWorker object at 0x7f62b45a2690>)
  File "/home/user1/alpa/examples/opt_serving/model/opt_model.py", line 785, in load_opt_params_worker_func
    wq = load_array(load_prefix + "self_attn.q_proj.weight")
  File "/home/user1/alpa/examples/opt_serving/model/opt_model.py", line 748, in load_array
    return np.load(os.path.join(path, key))
  File "/opt/conda/lib/python3.7/site-packages/numpy/lib/npyio.py", line 417, in load
    fid = stack.enter_context(open(os_fspath(file), "rb"))
FileNotFoundError: [Errno 2] No such file or directory: '/home/user1/alpa/examples/opt_serving/model/175B_resharded/numpy_weights/decoder.layers.0.self_attn.q_proj.weight'
2022-06-29 18:26:26,692	ERROR worker.py:95 -- Unhandled error (suppress with 'RAY_IGNORE_UNHANDLED_ERRORS=1'): ray::MeshHostWorker.run_executable() (pid=15307, ip=10.128.0.106, repr=<alpa.device_mesh.MeshHostWorker object at 0x7f62b45a2690>)
  File "/home/user1/alpa/alpa/device_mesh.py", line 253, in run_executable
    self.executables[uuid].execute_on_worker(*args, **kwargs)
  File "/home/user1/alpa/alpa/pipeline_parallel/pipeshard_executable.py", line 435, in execute_on_worker
    buffers[local_id] = self.global_buffers[global_id]
KeyError: 48
2022-06-29 18:26:26,698	ERROR worker.py:95 -- Unhandled error (suppress with 'RAY_IGNORE_UNHANDLED_ERRORS=1'): ray::MeshHostWorker.run_executable() (pid=15310, ip=10.128.0.106, repr=<alpa.device_mesh.MeshHostWorker object at 0x7f4e641bbd90>)
  File "/home/user1/alpa/alpa/device_mesh.py", line 253, in run_executable
    self.executables[uuid].execute_on_worker(*args, **kwargs)
  File "/home/user1/alpa/alpa/pipeline_parallel/pipeshard_executable.py", line 435, in execute_on_worker
    buffers[local_id] = self.global_buffers[global_id]
KeyError: 4176
Traceback (most recent call last):
  File "benchmark_text_gen.py", line 174, in <module>
    output_hidden_states=False)
  File "/opt/conda/lib/python3.7/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/transformers/generation_utils.py", line 1297, in generate
    **model_kwargs,
  File "/opt/conda/lib/python3.7/site-packages/transformers/generation_utils.py", line 1687, in greedy_search
    output_hidden_states=output_hidden_states,
  File "/home/user1/alpa/examples/opt_serving/model/wrapper.py", line 99, in __call__
    output_attentions=output_attentions)
  File "/home/user1/alpa/examples/opt_serving/model/wrapper.py", line 289, in inference_func
    logits_step = torch.from_numpy(np.array(output.logits)).to(device)
  File "/home/user1/alpa/alpa/device_mesh.py", line 1367, in __array__
    return np.asarray(self._value, dtype=dtype)
  File "/home/user1/alpa/alpa/device_mesh.py", line 1356, in _value
    for i in self.one_replica_buffer_indices
  File "/home/user1/alpa/alpa/device_mesh.py", line 1019, in get_remote_buffers
    return ray.get(obj_refs)
  File "/opt/conda/lib/python3.7/site-packages/ray/_private/client_mode_hook.py", line 105, in wrapper
    return func(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/ray/worker.py", line 1831, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(KeyError): ray::MeshHostWorker.get_buffers() (pid=15310, ip=10.128.0.106, repr=<alpa.device_mesh.MeshHostWorker object at 0x7f4e641bbd90>)
  File "/home/user1/alpa/alpa/device_mesh.py", line 210, in get_buffers
    return self.buffers[uuids]
KeyError: 12728
```

2. While trying to run the benchmark (benchmark_text_gen.py) using Ray, am getting some errors in Ray:

a. Unable to connect to GCS with ray start
4b No ray instance running, specify --address flag

Code:
```
ray start --head
python3 benchmark_text_gen.py --model alpa/opt-175b --path /home/usr1/alpa/examples/opt_serving/model/175B_resharded/ --debug
```


Please advise on these issues, I am using Jupyter-notebook in GCP , also could you provide any sample code Ray code, for OPT-175B. Thank you!

**Comment by zhisbug** [comment]
@BalajiAJ How many nodes are your class?

If on one node: you just need `ray start --head`.

If > 1 node: you need to start ray on every node and make sure ray processes report to the same head node. Try to follow this detailed guide https://docs.ray.io/en/latest/cluster/cloud.html#manual-cluster on how to manually start ray on a cluster of node.

Meanwhile, are all nodes able to access a shared NFS or the identical path shown in the log: `/home/usr1/alpa/examples/opt_serving/model/175B_resharded/`. You need to make sure all processes can see this path.

**Comment by BalajiAJ** [comment]
> @BalajiAJ How many nodes are your class?
> 
> If on one node: you just need `ray start --head`.
> 
> If > 1 node: you need to start ray on every node and make sure ray processes report to the same head node. Try to follow this detailed guide https://docs.ray.io/en/latest/cluster/cloud.html#manual-cluster on how to manually start ray on a cluster of node.
> 
> Meanwhile, are all nodes able to access a shared NFS or the identical path shown in the log: `/home/usr1/alpa/examples/opt_serving/model/175B_resharded/`. You need to make sure all processes can see this path.

@zhisbug oh! Not sure why I got a lesser count. The size of the consolidated single shard from OPT is 325.2 GB and there were no errors while consolidating the 992 shards into one.

![image](https://user-images.githubusercontent.com/19531638/176516290-850ec61a-ec82-4d40-9fdb-dcd9e522f5ff.png)

And, below script too ran without any errors
```
!python3 convert_to_numpy_weights.py --model alpa/opt-175b --path /home/usr1/metaseq/metaseq/scripts/reshard/ --debug
```

Could you please throw some light on what could have gone wrong in reduced numpy weight files?

And yes, its one node, ```ray start --head``` is the command I have run. Thanks for confirming 👍

**Comment by zhisbug** [comment]
Here is my guess:

OPT-175B has 96 layers, in my weights, each layer has 16 weights, so 16 * 96 = 1536 weights; Plus a `decoder.version`, and four weights corresponding to embedding and inputs -- in total 1536 + 5 = 1541.

In you case, assuming no bug in any script -- you have 1157 weights. Taking the 5 weights out gets 1157 - 5 = 1152 weights.
So each layer in your case has 1152 / 96  = 12 weights, meaning that you have 4 fewer weights than us.

I think the possible reason is your consolidation script does not split the `qkv` into `q`, `v`, `k`, `out_proj`.

looking at your screenshot above -- you indeed have qvk_proj as a combined version.

Which consolidation script did you use? Please take a look at #555 and try to use the step_1 script there (though I'll merge this branch soon)

**Comment by BalajiAJ** [comment]
> > @BalajiAJ Here is my guess:
> > OPT-175B has 96 layers, in my weights, each layer has 16 weights, so 16 * 96 = 1536 weights; Plus a `decoder.version`, and four weights corresponding to embedding and inputs -- in total 1536 + 5 = 1541.
> > In your case, assuming no bug in any script -- you have 1157 weights. Taking the 5 weights out gets 1157 - 5 = 1152 weights. So each layer in your case has 1152 / 96 = 12 weights, meaning that you have 4 fewer weights than us per layer.
> > I think the possible reason is your consolidation script does not split the `qkv` into `q`, `v`, `k`, `out_proj`.
> > looking at your screenshot above -- you indeed have qvk_proj as a combined version.
> > Which consolidation script did you use? Please take a look at #555 and try to use the step_1 script there (though I'll merge this branch soon)
> 
> @zhisbug I had used the [consolidate_fsdp_shards.py](https://github.com/facebookresearch/metaseq/tree/main/metaseq/scripts/consolidate_fsdp_shards.py) by following [OPT-175B Readme](https://github.com/facebookresearch/metaseq/blob/main/projects/OPT/download_opt175b.md) steps. Maybe that is the reason for reduction in weights. Anyways, I will follow the steps for consolidation from #555 as you mentioned and give it a try. Thanks!!

@zhisbug I used the step_1 script from #555 , but the script is not saving the consolidated checkpoint. I don't see any errors while executing the script, below is the snapshot. Please advise. Thank you.

![image](https://user-images.githubusercontent.com/19531638/176591624-1e07ca6e-c5b4-4b31-ae58-30d0f34546b1.png)

**Comment by zhisbug** [comment]
@BalajiAJ  I am fixing a few bugs and get #555 merged in about 30 mins

**Comment by zhisbug** [comment]
@BalajiAJ Just merged #555 
Also check out [this page](https://alpa-projects.github.io/tutorials/opt_serving.html) for a very detailed guide.

**Comment by wgimperial** [comment]
请问能否提供其他版本（如6.7b）的合并脚本，或者告诉我们如何修改原脚本。

**Comment by wgimperial** [comment]
Since we can't download OPT-2.7B weights file for alpa，
Is there any way to generate the weights file for alpa by myself？
I have download 2.7B and 6.7B  pretrained model weights files，how to merge file slices into one shard？ 
@zhisbug

**Comment by wgimperial** [comment]
The network is busy and the download time is too long.



| |
王刚
|
|
***@***.***
|




---- Replied Message ----
| From | Hao ***@***.***> |
| Date | 06/30/2022 22:23 |
| To | ***@***.***> |
| Cc | ***@***.******@***.***> |
| Subject | Re: [alpa-projects/alpa] Question - OPT 175B serving - Consolidated shards (Issue #554) |

@wgimperial could you let us know why you cannot download the 2.7b weights processed by us?

**Comment by zhisbug** [comment]
@wgimperial hello, I just updated the documentation with some new ways of downloading the consolidated singleton checkpoint for smaller versions of OPT. Please check #563 and [this new section](https://alpa-projects.github.io/tutorials/opt_serving.html#download-singleton) and see if it helps.

If you still have trouble downloading them, here are two ways:
- You can try to launch [this script](https://github.com/facebookresearch/metaseq/blob/main/metaseq/scripts/convert_to_singleton.py) from metaseq which can do the job by re-initializing the model in distributed GPUs and gather the weights. It might be buggy so you need to tune it to run correctly.

- Please join [Alpa slack](https://forms.gle/YEZTCrtZD6EAVNBQ7) and we might be able to transfer a version to you.

**Comment by zhisbug** [comment]
@BalajiAJ Things work well?

**Comment by wgimperial** [comment]
> @wgimperial hello, I just updated the documentation with some new ways of downloading the consolidated singleton checkpoint for smaller versions of OPT. Please check #563 and [this new section](https://alpa-projects.github.io/tutorials/opt_serving.html#download-singleton) and see if it helps.
> 
> If you still have trouble downloading them, here are two ways:
> 
> * You can try to launch [this script](https://github.com/facebookresearch/metaseq/blob/main/metaseq/scripts/convert_to_singleton.py) from metaseq which can do the job by re-initializing the model in distributed GPUs and gather the weights. It might be buggy so you need to tune it to run correctly.
> * Please join [Alpa slack](https://forms.gle/YEZTCrtZD6EAVNBQ7) and we might be able to transfer a version to you.

it's ok,thx

**Comment by BalajiAJ** [comment]
> @BalajiAJ Just merged #555 Also check out [this page](https://alpa-projects.github.io/tutorials/opt_serving.html) for a very detailed guide.
> 
> Fixed several bugs you encountered in the two new scripts.

@zhisbug Hi. Thanks for your help. I have ran the convert script to get numpy equivalent weights and got 1541 files as you mentioned above. Now while running the ray and benchmark script, I see the script is running, but I see an error related to CUDA (snapshot below). I have verified the CUDA, CUDNN versions and GPU info and everything looks fine. Any possible hints on this.

![image](https://user-images.githubusercontent.com/19531638/176960957-c2ce4846-1194-4342-809e-49013bf1bf32.png)
